### PR TITLE
Accurate Weather effects

### PR DESCRIPTION
**Pull request by fmatthew5876**
This PR is to add further improvements to weather emulation.

Fix: #2069 
Fix: #1737 

TBD:
- [x] Perfect rain emulation
- [x] Perfect snow emulation
- [x] Perfect fog emulation
- [x] Perfect sandstorm emulation

### CODE REVIEW

**Comment by fmatthew5876** [comment]
So far I've got accurate rain, snow, and fog. 

Sandstorm is almost there. I have everything figured out except how the initial acceleration values are calculated.

**Comment by fmatthew5876** [comment]
Now we finally have accurate rain, snow, fog, and sandstorm weather effects!

Since this is standalone and doesn't break games, I'd like to have it in 6.2 release if we can.

File: src/weather.cpp [hunk 1]
```diff
@@ -57,18 +59,22 @@ void Weather::Draw(Bitmap& dst) {
...
+static constexpr int num_sand_particles[] = { 128, 191, 255 };
 static constexpr int num_sand_colors = 4;
-static constexpr int num_sand_strength = 3;
 static constexpr auto sand_particle_rect = Rect{ 0, 0, 1, 2 };
 
 static constexpr auto sand_particle_bitmap_rect = Rect{
 	0, 0,
 	sand_particle_rect.width,
-	sand_particle_rect.height * num_sand_colors * num_sand_strength
+	sand_particle_rect.height * num_sand_colors,
```
Comment (Ghabry) [inline src/weather.cpp:77]: is this extra comma intentional? (I'm surprised that C++ accepts trailing commas)

File: src/game_screen.h [hunk 2]
```diff
@@ -87,11 +87,16 @@ class Game_Screen {
 	int GetWeatherStrength();
 
 	struct Particle {
+		int16_t t = 0;
 		int16_t x = 0;
 		int16_t y = 0;
-		int16_t life = 0;
-		int16_t speed = 0;
-		float angle = 0.0f;
+		// Thse are only used for sandstorm particles.
```
Comment (Ghabry) [inline src/game_screen.h:93]: these

File: src/weather.cpp (hunk 1 above)
Comment (fmatthew5876) [reply src/weather.cpp:77]: Not intentional, but modern C++ lets do you do this now.

**Comment by fmatthew5876** [comment]
> Is the movement speed of the Fog/Sand also random? In Player the left-movement is much slower than in RPG_RT. (tested multiple times)

I'll take a look at this in the debugger to verify fog values and algorithm

> 
> Imo the rain has on "lowest" the wrong colors. In RPG_RT almost all the rain is transparent but in Player you see more which are almost white.

Did you compare with true color RPG_RT? The colors are definately different without it, many instances the rain colors look more greenish on 16bit RPG_RT.

> 
> But without a direct comparison you would never notice this. Looks great!

**Comment by fmatthew5876** [comment]
Looking at rain in true color RPG_RT, it looks like you're right!

In fact there is another feature I missed entirely. The colors of the rain depend on the strength level!

| Strength | Color |
| -- | -- |
| 0 | 60 * t / 12 |
| 1 | 72 * t / 12 |
| 2 | 80 * t / 12 |

**Comment by fmatthew5876** [comment]
I did more analysis of rain and snow with RPG_RT true color, the pixel colors change based on strength. I also fixed the fog speeds and verified with x64dbg the fog movement algorithm matches RPG_RT.

This should be ready now

File: src/game_screen.h (hunk 2 above)
Comment (fmatthew5876) [reply src/game_screen.h]: Forgot this one. Fixed now.

**Pull request merged**